src/Compilers/Core/Portable/Operations/OperationWalker.cs

```diff
@@ -14,19 +12,14 @@ public abstract class OperationWalker : OperationVisitor
     {
         private int _recursionDepth;
 
-        internal void VisitArray<T>(IEnumerable<T> operations) where T : IOperation
+        internal void VisitArray(IOperation operation)
         {
-            foreach (var operation in operations)
+            foreach (var child in ((Operation)operation).ChildOperations)
```

Comment:
>((Operation)operation) [](start = 34, length = 22)

Is it really safe to assume that this public type always deals with nodes based on ```Operation``` type?

Reply:
Yes. `IOperation` is marked with our `InternalImplementationOnly` attribute, which will ensure our analyzers (that consumers get automatically when they reference our libraries) mark consumer implementation of these interfaces as errors. We also document that they are reserved for internal implementation only on every interface. Much like our symbol interfaces, if a consumer implements one themselves, there is no guarantee it will work with our APIs.

---
In reply to: [539664626](https://github.com/dotnet/roslyn/pull/49804#discussion_r539664626) [](ancestors = 539664626)

Reply:
> Yes. IOperation is marked with our InternalImplementationOnly attribute, which will ensure our analyzers (that consumers get automatically when they reference our libraries) mark consumer implementation of these interfaces as errors. We also document that they are reserved for internal implementation only on every interface. Much like our symbol interfaces, if a consumer implements one themselves, there is no guarantee it will work with our APIs.

We say pretty much the same for symbols, yet IDE turns around and implements their own. This looks like a breaking change to me. Consider adding this as an optimization when possible, rather than a complete replacement. 


---
In reply to: [539715866](https://github.com/dotnet/roslyn/pull/49804#discussion_r539715866) [](ancestors = 539715866,539664626)

Reply:
> Consider adding this as an optimization when possible, rather than a complete replacement. 

I did consider it, actually, but the branches did have a measurable impact on performance.

---
In reply to: [539718172](https://github.com/dotnet/roslyn/pull/49804#discussion_r539718172) [](ancestors = 539718172,539715866,539664626)

Reply:
> I did consider it, actually, but the branches did have a measurable impact on performance.

Then I think this should be approved by Compatibility Council.  

---
In reply to: [539726793](https://github.com/dotnet/roslyn/pull/49804#discussion_r539726793) [](ancestors = 539726793,539718172,539715866,539664626)

Reply:
Email setn.

---
In reply to: [539728958](https://github.com/dotnet/roslyn/pull/49804#discussion_r539728958) [](ancestors = 539728958,539726793,539718172,539715866,539664626)

Reply:
> Email sent.

Thanks, we can follow up later based on the decision, if needed. 

---
In reply to: [540550184](https://github.com/dotnet/roslyn/pull/49804#discussion_r540550184) [](ancestors = 540550184,539728958,539726793,539718172,539715866,539664626)